aten/src/ATen/autocast_mode.cpp

```diff
@@ -165,6 +177,35 @@ void set_autocast_cache_enabled(bool enabled) {
   cache_enabled = enabled;
 }
 
+
+// For gradient buffer stealing we will adjust the use count of tensors
+// which are persisted by cudagraphs, just as we need to adjust reference
+// count of tensors with hooks.
+size_t adjusted_use_count(const at::Tensor& t) {
+  auto use_count = t.use_count();
+  if (cudagraph_cached_tensorimpl_set_enabled) {
+    const std::lock_guard<std::mutex> lock(cudagraph_cached_tensorimpl_mutex);
+    use_count = use_count - cudagraph_cached_tensorimpl_set.count(t.unsafeGetTensorImpl());
+  }
+  return use_count;
+}
+
+void add_cudagraph_cached_tensor(const at::Tensor& t) {
+  TORCH_INTERNAL_ASSERT(cudagraph_cached_tensorimpl_set_enabled);
+  const std::lock_guard<std::mutex> lock(cudagraph_cached_tensorimpl_mutex);
+  cudagraph_cached_tensorimpl_set.emplace(t.unsafeGetTensorImpl(), weakref_type(t.getIntrusivePtr()));
+}
+
+void remove_cudagraph_cached_tensor(const at::Tensor& t) {
+  TORCH_INTERNAL_ASSERT(cudagraph_cached_tensorimpl_set_enabled);
+  const std::lock_guard<std::mutex> lock(cudagraph_cached_tensorimpl_mutex);
+  cudagraph_cached_tensorimpl_set.erase(t.unsafeGetTensorImpl());
+}
+
+void set_cudagraph_cached_tensors_enabled(bool enabled) {
+  cudagraph_cached_tensorimpl_set_enabled = enabled;
```

Comment:
without locks and without static?

Reply:
just as with autocast enabled, the global boolean will be set true before the map is written or read from, so you dont need a lock.

Reply:
I can add static... I saw that `cached_casts` did not use that, but maybe that's an omission ?